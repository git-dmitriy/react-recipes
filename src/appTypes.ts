export type CategoryItemTypes = {
  idCategory: string;
  strCategory: string;
  strCategoryThumb: string;
  strCategoryDescription: string;
};

export type MealItemTypes = {
  idMeal: string;
  strMeal: string;
  strMealThumb: string;
  strCategory?: string;
  strArea?: string;
  strInstructions?: string;
  strYoutube?: string;
};

export type RecipeItemTypes = {
  idMeal: string;
  strMeal: string;
  strMealThumb: string;
};

export type StateTypes = {
  favorites: MealItemTypes[] | [];
  theme: string | 'light' | 'dark';
  isLoading: boolean;
};

export type ActionTypes =
  | { type: ADD_TO_FAVORITE; payload: MealItemTypes }
  | { type: REMOVE_FROM_FAVORITE; payload: MealItemTypes }
  | { type: SWITCH_THEME; payload: 'light' | 'dark' }
  | { type: SET_IS_LOADING; payload: boolean };

type ADD_TO_FAVORITE = 'ADD_TO_FAVORITE';
type REMOVE_FROM_FAVORITE = 'REMOVE_FROM_FAVORITE';
type SWITCH_THEME = 'SWITCH_THEME';
type SET_IS_LOADING = 'SET_IS_LOADING';

export type ContextTypes = {
  state: StateTypes;
  addToFavorites: (item: MealItemTypes) => void;
  removeFromFavorites: (item: MealItemTypes) => void;
  switchTheme: (mode: 'light' | 'dark') => void;
  setIsLoading: (status: boolean) => void;
};
