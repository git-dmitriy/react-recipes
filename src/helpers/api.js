const API_KEY = '1';
const API_URL = `https://www.themealdb.com/api/json/v1/${API_KEY}`;

export const getMealById = async (mealId) => {
  const response = await fetch(API_URL + `/lookup.php?i=${mealId}`);
  return await response.json();
};

export const getMealByName = async (name) => {
  const response = await fetch(API_URL + `/search.php?f=${name}`);
  return await response.json();
};

export const getAllCategories = async () => {
  const response = await fetch(API_URL + '/categories.php');
  return await response.json();
};

export const getListOfCountries = async () => {
  const response = await fetch(API_URL + '/list.php?a=list');
  return await response.json();
};

export const getFilteredCategory = async (catName) => {
  const response = await fetch(API_URL + `/filter.php?c=${catName}`);
  return await response.json();
};

export const getFilteredCategoryByCountry = async (area) => {
  const response = await fetch(API_URL + `/filter.php?a=${area}`);
  return await response.json();
};

export const getRandomMeal = async () => {
  const response = await fetch(API_URL + 'random.php');
  return await response.json();
};
