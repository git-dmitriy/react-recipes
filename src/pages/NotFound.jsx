export const NotFound = () => {
  return <h1>404 Page is not found</h1>;
};
