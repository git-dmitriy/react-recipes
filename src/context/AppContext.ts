import { createContext } from 'react';
import { ContextTypes } from '@/appTypes';

export const AppContext = createContext({} as ContextTypes);
