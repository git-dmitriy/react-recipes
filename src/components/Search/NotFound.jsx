import { useEffect, useState } from 'react';
import { getRandomMeal } from '../../helpers/api';
import { Meal } from '../meals/Meal';

export const NotFound = ({ target }) => {
  const [randomMeal, setRandomMeal] = useState();
  useEffect(() => {
    getRandomMeal().then((data) => {
      setRandomMeal(data.meals[0]);
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div className='flex flex-col-reverse justify-around items-center my-6 px-2'>
      <ul className='rounded-lg overflow-hidden w-4/6 sm:w-1/2'>
        {randomMeal && <Meal {...randomMeal} />}
      </ul>
      <div className='w-full sm:w-1/2 pl-2 flex flex-col items-center text-center'>
        <h2 className=''>
          <span className='font-bold block'>Nothing found</span> for "{target}"
        </h2>
        <div>Try to cook this </div>
        <div className='w-12 h-1 bg-red-500 rounded mt-2 mb-4' />
      </div>
    </div>
  );
};
