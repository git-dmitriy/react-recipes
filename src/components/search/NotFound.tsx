import { useEffect, useState } from 'react';
import { getRandomMeal } from 'helpers/api';
import { Meal } from 'components/meals/Meal';

type P = {
  target: string;
};

export const NotFound: React.FC<P> = ({ target }) => {
  const [randomMeal, setRandomMeal] = useState();

  useEffect(() => {
    let cleanupFuse = true;
    getRandomMeal().then((data) => {
      cleanupFuse && setRandomMeal(data.meals[0]);
    });

    return () => {
      cleanupFuse = false;
    };
  }, []);

  return (
    <div className='flex flex-col-reverse justify-around items-center my-6 px-2'>
      <ul className='rounded-lg overflow-hidden w-4/6 sm:w-1/2'>
        {randomMeal && <Meal {...randomMeal} />}
      </ul>
      <div className='w-full sm:w-1/2 pl-2 flex flex-col items-center text-center'>
        <h2 className=''>
          <span className='font-bold block'>Nothing found</span> for "{target}"
        </h2>
        <div>Try to cook this </div>
        <div className='w-12 h-1 bg-red-500 rounded mt-2 mb-4' />
      </div>
    </div>
  );
};
