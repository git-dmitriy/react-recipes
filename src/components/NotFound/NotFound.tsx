import {useContext, useEffect, useState} from 'react';
import {getRandomMeal} from '@/api-utils.ts';
import {Meal} from '@components/Meal';
import {Layout} from '@components/Layout';
import {LostConnection} from '@components/LostConnection';
import {AppContext} from '@context/AppContext';
import {MealItemTypes} from '@/appTypes';


type P = {
    target: string;
};

export const NotFound: React.FC<P> = ({target}) => {
    const [randomMeal, setRandomMeal] = useState<MealItemTypes>();
    const [disconnected, setDisconnected] = useState(false);

    const {setIsLoading} = useContext(AppContext);

    useEffect(() => {
        let cleanupFuse = true;

        setIsLoading(true);

        getRandomMeal()
            .then((data) => {
                cleanupFuse && setRandomMeal(data.meals[0]);
            })
            .catch((e) => {
                console.warn(e);
                setDisconnected(true);
            })
            .finally(() => setIsLoading(false));

        return () => {
            cleanupFuse = false;
        };
    }, []);

    if (disconnected) {
        return (
            <Layout>
                <LostConnection/>
            </Layout>
        );
    }

    return (
        <div
            className='flex flex-col justify-around items-center my-6 px-2'
        >
            <div className='w-full sm:w-1/2 pl-2 flex flex-col items-center text-center'>
                <h2 className=''>
                    <span className='font-bold block'>Nothing found</span> for &quot;
                    {target}&quot;
                </h2>
                <div>Try to cook this</div>
                <div className='w-12 h-1 bg-red-500 rounded mt-2 mb-4'/>
            </div>
            <ul className='rounded-lg overflow-hidden w-4/6 sm:w-1/2'>
                {randomMeal && <Meal {...randomMeal} />}
            </ul>
        </div>
    );
};
