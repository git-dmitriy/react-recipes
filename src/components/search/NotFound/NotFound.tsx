import { useContext, useEffect, useState } from 'react';
import { getRandomMeal } from 'helpers/api';
import { Meal } from 'components/meals/Meals';
import { Layout } from 'components/layout/Layout';
import { LostConnection } from 'components/LostConnection';
import { AppContext } from 'context/AppContext';
import { MealItemTypes } from 'appTypes';

type P = {
  target: string;
};

export const NotFound: React.FC<P> = ({ target }) => {
  const [randomMeal, setRandomMeal] = useState<MealItemTypes>();
  const [disconnected, setDisconnected] = useState(false);

  const { setIsLoading } = useContext(AppContext);

  useEffect(() => {
    let cleanupFuse = true;

    setIsLoading(true);

    getRandomMeal()
      .then((data) => {
        cleanupFuse && setRandomMeal(data.meals[0]);
      })
      .catch((e) => {
        console.warn(e);
        setDisconnected(true);
      })
      .finally(() => setIsLoading(false));

    return () => {
      cleanupFuse = false;
    };
  }, []);

  if (disconnected) {
    return (
      <Layout>
        <LostConnection />
      </Layout>
    );
  }

  return (
    <div className='flex flex-col-reverse justify-around items-center my-6 px-2'>
      <ul className='rounded-lg overflow-hidden w-4/6 sm:w-1/2'>
        {randomMeal && <Meal {...randomMeal} />}
      </ul>
      <div className='w-full sm:w-1/2 pl-2 flex flex-col items-center text-center'>
        <h2 className=''>
          <span className='font-bold block'>Nothing found</span> for "{target}"
        </h2>
        <div>Try to cook this </div>
        <div className='w-12 h-1 bg-red-500 rounded mt-2 mb-4' />
      </div>
    </div>
  );
};
