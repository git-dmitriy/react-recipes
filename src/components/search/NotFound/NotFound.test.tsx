import { NotFound } from './NotFound';
import renderer from 'react-test-renderer';
import { AppContext } from 'context/AppContext';
import { ContextTypes } from 'appTypes';
import { MemoryRouter } from 'react-router-dom';
import { render, screen, waitFor } from '@testing-library/react';
import { getRandomMeal } from 'helpers/api';

jest.mock('helpers/api', (): { getRandomMeal: typeof getRandomMeal } => ({
  getRandomMeal: () =>
    Promise.resolve({
      meals: [
        {
          idMeal: '52942',
          strMeal: 'Roast fennel and aubergine paella',
          strCategory: 'Vegan',
          strArea: 'Spanish',
          strMealThumb:
            'https://www.themealdb.com/images/media/meals/1520081754.jpg',
        },
      ],
    }),
}));

const setIsLoading = jest.fn();
const context = {
  state: {
    favorites: [],
    theme: 'light',
    isLoading: false,
  },
  setIsLoading,
  addToFavorites: jest.fn(),
  removeFromFavorites: jest.fn(),
  switchTheme: jest.fn(),
};

it('should render correctly', () => {
  const target = '';

  const tree = renderer
    .create(
      <>
        <AppContext.Provider value={context as ContextTypes}>
          <MemoryRouter>
            <NotFound target={target} />
          </MemoryRouter>
        </AppContext.Provider>
      </>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('should load random meal', async () => {
  const target = 'Delicious dish';
  const meal = 'Roast fennel and aubergine paella';

  render(
    <AppContext.Provider value={context as ContextTypes}>
      <MemoryRouter>
        <NotFound target={target} />
      </MemoryRouter>
    </AppContext.Provider>
  );

  const searchMeal = screen.getByRole('heading');

  expect(searchMeal).toBeInTheDocument();

  await waitFor(() => {
    const randomMealTitle = screen.getByText(meal);

    expect(randomMealTitle).toBeInTheDocument();
  });
});
