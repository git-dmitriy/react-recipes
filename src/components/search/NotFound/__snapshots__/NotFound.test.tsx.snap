// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="flex flex-col justify-around items-center my-6 px-2"
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <div
    className="w-full sm:w-1/2 pl-2 flex flex-col items-center text-center"
  >
    <h2
      className=""
    >
      <span
        className="font-bold block"
      >
        Nothing found
      </span>
       for "
      
      "
    </h2>
    <div>
      Try to cook this 
    </div>
    <div
      className="w-12 h-1 bg-red-500 rounded mt-2 mb-4"
    />
  </div>
  <ul
    className="rounded-lg overflow-hidden w-4/6 sm:w-1/2"
  />
</div>
`;
